### ISSUE TITLE
surf::Exception does not implement std::error::Error

### ISSUE THREAD

**Issue by dtolnay** [http-rs/surf#86]
This is an unfriendly choice for errors returned by a library because it makes them not work with `?`.

```rust
async fn repro() -> Result<(), failure::Error> {
    let _ = surf::get("https://www.rust-lang.org").await?; // doesn't work
    Ok(())
}
```

```rust
async fn repro() -> anyhow::Result<()> {
    let _ = surf::get("https://www.rust-lang.org").await?; // doesn't work
    Ok(())
}
```

Application-focused error types like failure::Error are built on `impl<T: std::error::Error> From<T>` which is why it matters that library errors implement std::error::Error.

**Comment by yoshuawuyts** [comment]
@dtolnay thanks for pointing this out; I agree this is not right, and we should fix this. Thanks for reporting, and sorry this spilled over into your issue tracker!

**Comment by matklad** [comment]
@yoshuawuyts this might want to be marked with 2.0 milestone.

**Comment by derekdreery** [comment]
From `anyhow::Error` docs

> Error works a lot like Box<dyn std::error::Error>, but with these differences:
> - Error requires that the error is Send, Sync, and 'static.
> - Error guarantees that a backtrace is available, even if the underlying error type does not provide one.
> - Error is represented as a narrow pointer — exactly one word in size instead of two.

 - The first one of these is fixed by using `Box<dyn std::error::Error + Send + Sync>` (I'm informed boxed errors are `'static` unless they have a lifetime parameter).
 - `anyhow::Error` creates a backtrace where the error is first converted into an `anyhow::Error`, I'm not sure that adds much in this use case over not having one.
 - `anyhow::Error` is a thin pointer — so you add one indirection to save one usize width. Again I don't think this matters much in this context: your `Result<T>` will be sized by `T` (unless it is less than 2 `usize`s), and unless you've only got a tiny amount of memory (in which case you can't use `alloc`, and therefore this crate) the nature of making http requests means that you're not going to create enough to make memory usage an issue.

All in all just use `Box<dyn std::error::Error + Send + Sync>` which is also `'static`. 1 less dependency and does the job of a type-erased error.

I made an RFC for [adding the type alias to std](https://github.com/rust-lang/rfcs/pull/2820).

**Comment by dtolnay** [comment]
@derekdreery `Box<dyn Error + Send + Sync>` does not implement std::error::Error so it does not address this issue at all.

**Comment by matklad** [comment]
>I'm informed boxed errors are 'static unless they have a lifetime parameter

The precise rule here is that lifetime elision works for trait object (`dyn Trait`), and elided lifetime is `'static`, *unless* the trait object is behind an `&[mut] 'a`, in which case the lifetime is `'a`. That is, boxed errors are not special, a DST `dyn Error` is also equivalent to `dyn Error + 'static`.

>All in all just use Box<dyn std::error::Error + Send + Sync> which is also 'static. 1 less dependency and does the job of a type-erased error.

The general problem here is that any two type-erased errors are not interoperable with each other, be it `Box<dyn Error>`, `anyhow::Error` or `failure::Error`.

In the ideal world, `anyhow::Error` would contain `From<Box<dyn Error>> for anyhow::Error` impl, but that is incoherent, because upstream crate (std) could add `impl Error for Box<dyn Error>`. The irony is that std can't actually add this imp, because that would make the two `From` impls for `Box<dyn Error>` overlap. 

It looks like, if a *library* wants to use dynamic errors internally, it has to do some manual wrapping into `struct Error { dyn_error: Box<dyn std::error::Error> }` at the boundary.

**Comment by derekdreery** [comment]
@dtolnay  you're right about it not implementing `std::error::Error`, but in my experience it seems to work with `?`. I don't know for sure if this is the case, and if so why.

**Comment by zkat** [comment]
Can we uhhhh....

use [`thiserror`](https://crates.io/crates/thiserror) to wrap every single error that `surf` can return, since this is a library? I'd be willing to put together a PR to do that, I think. Or at least try. `surf` really shouldn't be using either `anyhow` nor `failure` for this sort of thing.

**Comment by Fishrock123** [comment]
I believe this was effectively solved by moving to [`http_types::Error`](https://docs.rs/http-types/2.3.0/http_types/struct.Error.html) in https://github.com/http-rs/surf/commit/c8366fb2b9d7d59af8e9cc7a5f9d2173d459e7ec. cc @yoshuawuyts

**Comment by dtolnay** [comment]
The `http_types::Error` type in your link does not implement std::error::Error so I'm not sure that would solve it.

**Comment by joshtriplett** [comment]
This seems to still be an issue today. I tried surf 2.0.0-alpha.4, and attempted to use `?` in a function using `anyhow::Error` as its error type. This failed because `surf::Error` doesn't implement `std::error::Error`.

**Comment by Fishrock123** [comment]
I don't think this is going to make the cut for 2.0 as I am unsure what to do about this.

**Comment by leo60228** [comment]
Fixing this requires a breaking change. I think this should've been considered a release blocker.

**Comment by Fishrock123** [comment]
We will have a 3.0 at some point and we should think about it for that.

**Comment by jbr** [comment]
I'm pretty sure that the type system won't let us implement std::error::Error for http_types::Error, but we can make it easier to use http_types::Errors inside of anyhow, and possibly even unwrap them into anyhow::Errors outside of surf, as the status_code isn't important once it's outside of a surf context. This is at least partly a http_types issue

**Comment by Fishrock123** [comment]
That is also my understanding, however the error handling project group is also working on fixing `StdError` in ways that should help, as far as I'm aware.

**Comment by ohmree** [comment]
A relatively easy fix for users of eyre (which is a fork of anyhow, so I believe it should also work for the latter) is:

```rust
surf::get(my_url).await.map_err(|e| eyre!(e))?
// or
surf::get(my_url).await.map_err(|e| anyhow!(e))?
```

Alternatively you can also return a `eyre::Result<T, surf::Error>` (or `anyhow::Result<T, surf::Error>`) from your functions, although (with eyre at least) you miss out on pretty output that shows where the error occured and possibly more info (depends on the edition of eyre you use).

**Comment by lefuturiste** [comment]
Is this issue on the roadmap?
